Nodejs/Product/Nodejs/SharedProject/HierarchyIdMap.cs

```diff
@@ -57,33 +47,21 @@ public void Remove(HierarchyNode node)
         {
             VisualStudio.Shell.ThreadHelper.ThrowIfNotOnUIThread();
 
-            if (node == null)
-            {
-                throw new ArgumentNullException(nameof(node));
-            }
+            Debug.Assert(node != null, "Called with null node");
 
-            lock (this.theLock)
-            {
-                var i = (int)node.ID - 1;
-                if (0 > i || i >= this.ids.Count)
-                {
-                    throw new InvalidOperationException($"Invalid id. {i}");
-                }
+            var idx = (int)node.ID - 1;
 
-                var weakRef = this.ids[i];
-                if (weakRef == null)
-                {
-                    throw new InvalidOperationException("Trying to retrieve a node before adding.");
-                }
+            var removeCheck = this.nodes.TryGetValue(idx, out var weakRef);
 
-                if (weakRef.TryGetTarget(out var found) && !object.ReferenceEquals(node, found))
-                {
-                    throw new InvalidOperationException("The node has the wrong id.");
-                }
+            Debug.Assert(removeCheck, "How did we get an id, which we haven't seen before");
+            Debug.Assert(weakRef != null, "Double delete is not expected.");
 
-                this.ids[i] = null;
-                this.freedIds.Push(i);
-            }
+            var checkReference = weakRef.TryGetTarget(out var found) && object.ReferenceEquals(node, found);
+
+            Debug.Assert(checkReference, "The node has the wrong id.");
+
+            this.nodes[idx] = null;
```

Comment:
>nodes [](start = 17, length = 5)

It seems like it might be safer to call `TryRemove` (or `TryUpdate` if we want to keep the empty slot, for some reason).

Reply:
yes